src/Adapter/Discount/Update/DiscountDuplicator.php

```diff
@@ -0,0 +1,331 @@
+use PrestaShop\PrestaShop\Core\Hook\HookDispatcherInterface;
+use PrestaShop\PrestaShop\Core\Repository\AbstractObjectModelRepository;
+use PrestaShop\PrestaShop\Core\Util\String\StringModifierInterface;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+/**
+ * Duplicates discount
+ */
+class DiscountDuplicator extends AbstractObjectModelRepository
+{
+    public function __construct(
+        protected readonly DiscountRepository $discountRepository,
+        protected readonly HookDispatcherInterface $hookDispatcher,
+        protected readonly Connection $connection,
+        protected readonly string $dbPrefix,
+        protected readonly TranslatorInterface $translator,
+        protected readonly StringModifierInterface $stringModifier,
+    ) {
+    }
+
+    /**
+     * Global process of duplicating a discount
+     */
+    public function duplicate(DiscountId $discountId): DiscountId
+    {
+        $oldDiscountId = $discountId->getValue();
+        $this->hookDispatcher->dispatchWithParameters(
+            'actionAdminDuplicateBefore',
+            ['id_discount' => $oldDiscountId]
+        );
+        $newDiscountId = $this->duplicateDiscount($discountId);
+
+        $this->duplicateRelations($discountId, $newDiscountId);
+
+        $this->hookDispatcher->dispatchWithParameters(
+            'actionAdminDuplicateAfter',
+            ['id_discount' => $oldDiscountId, 'id_discount_new' => $newDiscountId->getValue()]
+        );
+
+        return $newDiscountId;
+    }
+
+    /**
+     * Duplicates the discount itself
+     */
+    private function duplicateDiscount(DiscountId $sourceDiscountId): DiscountId
+    {
+        $cartRule = $this->discountRepository->get($sourceDiscountId);
+        $duplicatedCartRule = $cartRule->duplicateObject();
+
+        if (!$duplicatedCartRule instanceof CartRule) {
+            throw new CannotDuplicateDiscountException('Failed to duplicate discount');
+        }
+
+        $duplicatedCartRule->name = $this->getNewProductName($cartRule->name);
+
+        if ($duplicatedCartRule->code) {
+            $duplicatedCartRule->code = $this->generateRandomCode();
+        }
+
+        $duplicatedCartRule->date_add = date('Y-m-d H:i:s');
+        $duplicatedCartRule->active = false;
+        $duplicatedCartRule->update();
+
+        return new DiscountId((int) $duplicatedCartRule->id);
+    }
+
+    /**
+     * Provides duplicated discount name
+     *
+     * @param array<int, string> $oldDiscountLocalizedNames
+     *
+     * @return array<int, string>
+     */
+    private function getNewProductName(array $oldDiscountLocalizedNames): array
+    {
+        $newLocalizedNames = [];
+        foreach ($oldDiscountLocalizedNames as $langId => $oldName) {
+            $langId = (int) $langId;
+            $namePattern = $this->translator->trans('copy of %s', [], 'Admin.Catalog.Feature', Language::getLocaleById($langId));
+            $newName = sprintf($namePattern, $oldName);
```

Comment:
```suggestion
            $newName = $this->translator->trans('copy of %s', [$oldName], 'Admin.Catalog.Feature', Language::getLocaleById($langId));
```
or maybe
```suggestion
            $newName = $this->translator->trans('copy of %s', ['%s' => $oldName], 'Admin.Catalog.Feature', Language::getLocaleById($langId));
```

But `sprintf` is redundant since `trans` already handles placeholders